Wrapper for Results

henrij22 opened the pull request:
the wrapper exports the type of the stored result and is able to return the result as a vector or a matrix

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,146 @@
+  {
+    template <typename Derived>
+    auto operator()(const Eigen::MatrixBase<Derived>& vec) const {
+      return fromVoigt(vec.derived(), false);
```
Can you maybe generalize this, to also alllow for true and false?

henrij22 replied:
Yeah the `asVec` function can take a second argument i think

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,146 @@
+      return value_;
+  }
+
+  void emplace(const InputType& t) { value_ = std::move(t); }
```
This is not how move works. https://godbolt.org/z/8ezEPer6j The move here does nothing since it cannot move from a constant reference.

henrij22 replied:
Thkx !

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,146 @@
+  REGISTER_RESULTTYPE(vectorPotential, worldDim, 1);
+  REGISTER_RESULTTYPE(divergenceOfVectorPotential, 1, 1);
+
+  REGISTER_RESULTTYPE(BField, 1, 1);
```
Both of them are vectors of worldDim (HField and BField)

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,146 @@
+} // namespace ResultType
+
+template <template <typename, int, int> class RT1, template <typename, int, int> class RT2>
+constexpr bool isSameResultType = std::is_same_v<RT1<double, 1, 1>, RT2<double, 1, 1>>;
```
If you are wondering how to do this more general: https://stackoverflow.com/a/52322078/12484740.

typename... but has to be `typename, int, int`,  there is no really general way to do this

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,146 @@
+  static constexpr bool registeredTypeIsVector = Traits::IsVectorAtCompileTime;
+
+public:
+  using VecType    = std::conditional_t<registeredTypeIsDynamic, Eigen::MatrixX<ScalarType>,
```
Why not just use invoke result? It should als just work with the dynamic stuff. Maybe there is some problem if it doesn't work

henrij22 replied:
Yeah i tried that initially, but it somehow deduced some kind of static matrix from that, thats why I did that. Also I think its okay that `vec` and `mat` for the custom type are just again `MatrixXd`s

rath3t replied:
I agree but I think the code would simply, if we generalize this. Then, the use can also do more fancy types like like our initial types that are max 9x9 but can also be smaller. So basically, we should just support any Eigen::Matrix<....>

henrij22 replied:
Can you make an example on how to do that?

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,146 @@
+  using ResultType = RT;
+
+  const InputType& operator()() const { return value_; }
+  const InputType& operator()() { return value_; }
```
The pair of these operators does not make sense.

The second one has to be a mutable reference
```cpp
const InputType& operator()() const { return value_; }
  InputType& operator()() { return value_; }

rath3t replied:
But I wonder if this is really needed. The first one may be sufficent

henrij22 committed:
first idea

change resulttype to structs

meta programming

Add Voigt parameter + tests

work and fix

fix python

fix kl and doc

fix again

add concept

wip discussion

Working in ResultTypeContainer for linearelastic

nonlinear + test cases

fix tests

fix python and custom type

fix python kl test

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,146 @@
+
+private:
+  InputType value_{};
+  typename RT::Vectorizer vectorizer_{};
```
Since these operators are stateless, they dont havew to be member variables.
This is even wasteful sicne the standard dicates that every member variables hast to have an adress.
So even if the types are empty they waste memory. For the later case this can be fixed with https://en.cppreference.com/w/cpp/language/attributes/no_unique_address. But since there is not a constructor why gets the vectorizer the construction here makes no since anyway.

rath3t commented on ikarus/finiteelements/mechanics/enhancedassumedstrains.hh:
```diff
@@ -158,43 +159,41 @@ forward the
-    return resultVector;
+    if constexpr (isSameResultType<RT, ResultType::linearStress>) {
+      if (isDisplacementBased())
+        return DisplacementBasedElement::template calculateAt<RT>(req, local);
```
If the eleemnt is displacementbased we shoudl forward all calls of calculateAt the DisplacementBasedElement.

rath3t replied:
Not only if RT == ResultType::linearStress

henrij22 replied:
Ah yeah the linter in clion is highliting it, but it works without the typename

rath3t commented on ikarus/finiteelements/mechanics/enhancedassumedstrains.hh:
```diff
@@ -298,7 +297,7 @@ protected:
 
 private:
-  using EASVariant = EAS::Variants<Geometry>::type;
+  using EASVariant = typename EAS::Variants<Geometry>::type;
```
In c++ 20 the paper https://wg21.link/P0634R3 got accepted which makes typename obsolete in some contexts

rath3t replied:
See the exaample with using therein

henrij22 replied:
Can you point me to the example? I can not see it
Also why did I change that line :D

rath3t replied:
I can not create a link to the example but therin search for `using Ptr = PtrTraits<T>::Ptr;`

henrij22 committed:
fixes from code review

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,156 @@
+namespace Ikarus {
+namespace Util {
+
+  struct VectorizeWithVoigt
```
Maybe we put these in an Impl namespace since they are an implementation detail.

rath3t commented on ikarus/python/finiteelements/nonlinearelastic.hh:
```diff
@@ -104,9 +104,9 @@ void registerNonLinearElastic(pybind11::handle scope, pybind11::class_<NonLinear
-         ResultType resType) {
-        if (resType == ResultType::PK2Stress)
-          return self.template calculateAt<ResultType::PK2Stress>(req, local);
+         std::string& resType) {
```
This should by value or const reference. Mutablwe reference does not make sense here.

rath3t commented on ikarus/python/test/linearelastictest.py:
```diff
@@ -125,14 +125,10 @@ def fixLeftHandEdge(vec, localIndex, localView, intersection):
-        lambda e, x: fes[indexSet.index(e)].calculateAt(
-            req, x, iks.ResultType.linearStress
-        )[0]
+        lambda e, x: fes[indexSet.index(e)].calculateAt(req, x, "linearStress")[0]
```
This has to be also a MultiDimensionCodimWrapper at some point. Can you create an issue, for this.

henrij22 replied:
Why do we need that?

rath3t replied:
Because the indices of the elements are non-unique for every geometry type in the indexSets.

rath3t replied:
So if you have three triangles and two quads, you would get:

```cpp
indexSet.index(triangle1) -> 0
indexSet.index(triangle2) -> 1
indexSet.index(triangle3) -> 2
indexSet.index(quad1) -> 0
indexSet.index(quad2) -> 1
```
So all this breaks if your grid has elements with different geometry types.

rath3t replied:
The MultiDimensionCodimWrapper takes care of this and would return something like 

```cpp
mdcm.index(triangle1) -> 0
mdcm.index(triangle2) -> 1
mdcm.index(triangle3) -> 2
mdcm.index(quad1) -> 3
mdcm.index(quad2) -> 4
```

This also breaks in the result function currently.

henrij22 replied:
Ah yes thats true, I remember our discussion about that

rath3t commented on ikarus/utils/functionhelper.hh:
```diff
@@ -37,5 +39,25 @@ void obtainLagrangeNodePositions(const LV& localView,
+  for (auto i : std::views::iota(0, numberOfVertices))
+    positions.push_back(referenceElement.position(i, dim));
+
+  return positions;
```
This should return a range view instead of a std::vector. Then it will not allocate an std::vector all the time. You can pass the iota view to an transformed_view and return that. https://en.cppreference.com/w/cpp/ranges/transform_view

rath3t replied:
And the function should be named different. First with the get and the it is still misleading, since I may return the global positions of the vertices. Maybe somethign like `referenceElementVertexPositions`

henrij22 replied:
Nice, added

rath3t commented on tests/src/resultcollection.hh:
```diff
@@ -172,5 +159,20 @@ inline auto linearStressResultsOfTetrahedron = []<typename NOP, typename FE>(NOP
+auto stressResultsToMatrix(const CompileTimeMatrix& expectedResults) {
+  constexpr int vertices   = CompileTimeMatrix::CompileTimeTraits::RowsAtCompileTime;
+  constexpr int voigtComps = CompileTimeMatrix::CompileTimeTraits::ColsAtCompileTime;
+  constexpr int matrixSize = (-1 + Ikarus::ct_sqrt(1 + 8 * vertices)) / 2;
```
This line seems strange. Why is the number of vertices  used here so complicated. The number of vertices os a compile time constant therfore I assume this is restricted to one geometry type.

henrij22 replied:
This is just bad wording, The `vertices` are the amount of rows, and as I supply the results as row Vectors, the amount of vertices therefore is deduced from the matrix und therefore its generally usable
This function is just used to transform the results in the `resultCollection.hh`

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -236,29 +232,42 @@ template <typename NonLinearOperator>
+                             << expectedResult << "\n"
+                             << messageIfFailed;
+  } else
+    t.check(false) << "Element can not provide the requested RsultType " << Ikarus::toString<resType>()
```
Failing at runtime maybe does not make sense here. I imagine someone writes a new element and want to perform checks on it. Then this function is used in the tests. Then the developer could be warned that this check can not be performed for the new element at compile time, do tom some reason.

rath3t commented on tests/src/testferesulttypes.cpp:
```diff
@@ -0,0 +1,135 @@
+auto testRTs() {
+  TestSuite t("Test FE ResultTypes");
+
+  auto testStuff = [&]<ElementHasResultType rt, bool asVec>(Shape expectedShapeVec, Shape expectedShapeMat) {
```
Better lambda name!

rath3t commented on ikarus/io/resultfunction.hh:
```diff
@@ -97,10 +97,10 @@ public:
-    if constexpr (std::is_same_v<UserFunction, Impl::DefaultUserFunction>)
-      return toString(resType);
-    else
+    if constexpr (requires { UserFunction::name(); })
```
Here you require that `UserFunction` has a static method name `UserFunction::name()`, but below you use a normal member function `obj.name()`. You should change this require check using the actual object, or also call the name function below as a static function without object `UserFunction::name()`

henrij22 replied:
Ah good point. 
I would only require a normal member function, then the interface is less restrictive. 

Also I made the change because I think a ResultEvaluator doesn't need to have a 'name' function, as it could fall back to the name of the result 

I should also add a Concept for ResultEvaluators

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -236,29 +232,42 @@ template <typename NonLinearOperator>
 
-template <Ikarus::ResultType resType>
-[[nodiscard]] auto checkCalculateAt(auto& nonLinearOperator, auto& fe, const auto& feRequirements,
+template <template <typename, int, int> class resType, bool voigt = true>
```
Maybe change the bool to a different name such as vectorizedResult or something

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,156 @@
+  using InputType  = std::conditional_t<inputIsVec, VecType, RegisteredType>;
+  using ResultType = RT;
+
+  const InputType& operator()() const { return value_; }
```
Do we really need that function. Maybe its better to remove this an explicitly use `asVec` or `asMat`.

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,156 @@
+
+  explicit ResultTypeContainer() = default;
+
+  void emplace(const InputType&& value) { value_ = std::move(value); }
```
const forwarding reference does also not make sense this has to be mutable. Anyway I would remove this emplace function and just create this via the appropriate constructors for ResultTypeContainer. And maybe think about renaming to ResultContainer or ResultWrapper.

rath3t committed:
draft refactor from review

henrij22 committed:
Fix Merge Conflict

henrij22 commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,189 @@
+  REGISTER_RESULTTYPE_IMPL(structName, rowsExpr, ColsExpr, rowsExpr, ColsExpr,           \
+                           Ikarus::Impl::VectorizeWithVoigt<strainlike>, Ikarus::Impl::MatricizeWithVoigt<strainlike>)
+
+#define REGISTER_RESULTTYPE(structName, rowsExpr, ColsExpr, MaxRowsExpr, MaxColsExpr)      \
```
I think this is more or less ready from my side. 
We should also add some bits of documentation to these Macros on how they can be used. 
Also is `simple` a good name here?

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,189 @@
+/**
+ *
+ * \ingroup FEParameterTags
+ * \brief A strongly typed enum class representing the type of the result request
```
Outdated comment. Can you make this like 
```cpp
/**
 * \file ferequirements.hh
 * \brief Definition of the LinearElastic class for finite element mechanics computations.
 * \ingroup finiteelements
 */
```

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,189 @@
+                                       const ResultTypeValueType&>;
+  using MatType =
+      std::invoke_result_t<decltype(&RT::Matricizer::template transform<VecType, rowsAtCompileTime, colsAtCompileTime>),
+                           const VecType&, int, int>;
```
Very ugly clang format here :/

rath3t commented on ikarus/finiteelements/mechanics/kirchhoffloveshell.hh:
```diff
@@ -189,14 +189,25 @@ public:
+   * \tparam RT The type representing the requested result.
    */
-  template <ResultType resType>
+  template <template <typename, int, int> class RT>
```
We can also disable this function completly if the result type is not avaibable. Basically in all element add `requires canProvideResultType<RT>`

henrij22 replied:
Then the static assert is not needed in the function

rath3t replied:
Here i don't see a static_Assert only the throw. and yes this would not be needed.

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,189 @@
+ * \tparam RT A specified ResultType
+ * \tparam storedValueIsVector boolean indicating whether the stored result is in its vector form, defaults to true
+ */
+template <typename RT, bool storedValueIsVector = true>
```
instead of `bool` we can use an enum ResultShape with `Matrixshape`and vectorshape` or something better then below at call site, there is not this magic `true` inside calculateAt, which makes it more readable.

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,189 @@
+  /**
+   * \brief Returns the stored value as Vector
+   */
+  auto asVec() const {
```
Better doxygen comments in this class

henrij22 replied:
should we use `decltyp(auto)` for return type so we can return by reference?

rath3t replied:
I think this would still return `value_` by value. Since this is a small you can also return by value except if the result type is a big matrix. Then you have to refactor like https://godbolt.org/z/9G77E1jfo.

rath3t commented on ikarus/finiteelements/mechanics/linearelastic.hh:
```diff
@@ -210,19 +210,30 @@ public:
+   * \return boolean indicating if a requested result can be provided
+   */
+  template <template <typename, int, int> class RT>
+  static constexpr bool canProvideResultType() {
```
This function should not only be constexpr but consteval maybe

rath3t commented on ikarus/io/resultfunction.hh:
```diff
@@ -82,7 +82,7 @@ public:
       Dune::FieldVector<ctype, griddim> val(0.0);
 
-      auto sigma = fes_->at(0).template calculateAt<resType>(feRequirements_, val);
+      auto sigma = fes_->at(0).template calculateAt<resType>(feRequirements_, val).asVec();
```
If `asVec` is hard coded here, then the `return static_cast<int>(sigma.rows() * sigma.cols());` can be simplified

henrij22 replied:
Yeah I would for now leave it hardcoded

rath3t replied:
Ah no wait this function is now not even needed anymore. You can calculate the vector size at compile time

henrij22 replied:
But not if we have dymaic size result types right?

rath3t replied:
Ah yes you are right.

rath3t commented on ikarus/python/finiteelements/linearelastic.hh:
```diff
@@ -44,6 +44,20 @@ namespace Ikarus::Python {
+      [](LinearElastic& self, const FERequirements& req, const Dune::FieldVector<double, Traits::mydim>& local,
+         std::string resType) {
+        if (resType == "linearStress")
+          return self.template calculateAt<ResultType::linearStress>(req, local).asVec();
```
This should be a function `registerCalculateAt`, which accepts the result types and creates the registration. Then all the element register the function by calling

```cpp
registerCalculateAt(scope,cls,std::make_tuple(linearStress,....));
```

rath3t replied:
You can again iterate over the tuple with Hybrid:forEach and if the string matches one of `toString(RT)`, then you can return the corresponding value. Note that, since calculateAt is called with the string the result type is only know at runtime, therefore in the end all the result have to be converted to `Eigen::VectorXd` or directly to `py::array_t<double>`.

henrij22 replied:
Okay yeah ah function is good for that, but how then I need a list somewhere of all `RT`s right? 
And i cant iterate over the namespace ... We don't have reflection :D

rath3t replied:
For now you can pass that list by hand for each element as in `registerCalculateAt(scope,cls,std::make_tuple(linearStress,....));` but the elements can also export exaclty this tuple that they support.

henrij22 replied:
But i still have to match the templates to the strings somehow?

rath3t replied:
I think thats what i try to indicate above
First we may use https://gitlab.dune-project.org/core/dune-common/-/blob/master/dune/common/tuplevector.hh instead of std::tuple since it has the convenient operator[]
```cpp
template<Element>
void registerCalculateAt(scope,cls,tuple)
{
  using Traits         = typename Element::Traits;
  using FERequirements = typename Element::FERequirementType;
  cls.def(
      "calculateAt",
      [](Element& self, const FERequirements& req, const Dune::FieldVector<double, Traits::mydim>& local,
         std::string resType) {
        Dune::Hypbrid::forEach(tuple,[&](auto i){
        if (resType == toString(tuple[i]))
          return Eigen::VectorXd(self.template calculateAt<tuple[i]>(req, local).asVec());
});
          DUNE_THROW(Dune::NotImplemented, "The requested result type is not supported by the given element");
      },
      pybind11::arg("feRequirements"), pybind11::arg("local"), pybind11::arg("resultType"));


});
}

rath3t replied:
You can even create a string of supported types which can be used in the error message:
```cpp
std::string restypes;
Dune::Hypbrid::forEach(tuple,[&](auto i){
restypes = restypes + toString(tuple[i]) + ",";
});
```

henrij22 replied:
Okay it works now, but it got a bit ugly, as i was not able to return out of the lamba. 
Please have a look

rath3t commented on ikarus/python/finiteelements/nonlinearelastic.hh:
```diff
@@ -104,9 +104,9 @@ void registerNonLinearElastic(pybind11::handle scope, pybind11::class_<NonLinear
-        if (resType == ResultType::PK2Stress)
-          return self.template calculateAt<ResultType::PK2Stress>(req, local);
+         std::string resType) {
+        if (resType == "PK2Stress")
```
Same as above

rath3t commented on ikarus/utils/functionhelper.hh:
```diff
@@ -37,5 +39,22 @@ void obtainLagrangeNodePositions(const LV& localView,
+ * \param fe finite element
+ * \return std::vector of LocalCoordinates of the element
+ */
+template <typename FiniteElement>
```
FE instead of FiniteElement

rath3t commented on ikarus/utils/functionhelper.hh:
```diff
@@ -37,5 +39,22 @@ void obtainLagrangeNodePositions(const LV& localView,
+ * \return std::vector of LocalCoordinates of the element
+ */
+template <typename FiniteElement>
+auto referenceElementVertexPositions(FiniteElement& fe) {
```
This can be generalized to `referenceElementSubEntitiyPositions` and the `dim`argument is passed as `codim`argument.

rath3t commented on tests/src/testcommon.hh:
```diff
@@ -236,29 +233,41 @@ template <typename NonLinearOperator>
+        computedResults.row(i++) = result.transpose();
+      } else {
+        auto result = fe.template calculateAt<resType>(feRequirements, pos).asMat();
+        Eigen::Map<Eigen::VectorXd> vector(result.data(), result.size());
```
Why this inderiction couldn't you do `computedResults.row(i++) = result.reshaped().transpose();`

rath3t commented on tests/src/testnonlinearelasticity.hh:
```diff
@@ -157,6 +157,8 @@ auto NonLinearElasticityLoadControlNRandTR(const Material& mat) {
                                                                   << maxDisp;
   }
 
+  auto result = fes.front().template calculateAt<ResultType::PK2Stress>(req, {0, 0}).asVec();
```
This value is not used. Since it is auto not even the expected type is checked.

rath3t reviewed (changes requested):
This is going in the right direction have a look at my comments. Add also add comments to the macros

rath3t commented on ikarus/io/resultfunction.hh:
```diff
@@ -45,7 +45,7 @@ namespace Impl {
 DefaultUserFunction)
  */
-template <typename FE, ResultType resType, typename UserFunction = Impl::DefaultUserFunction>
+template <typename FE, template <typename, int, int> class resType, typename UserFunction = Impl::DefaultUserFunction>
```
This should be named RT not resType

rath3t replied:
From RT you can also get the size of the result.

henrij22 commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,189 @@
+  REGISTER_RESULTTYPE_IMPL(structName, rowsExpr, ColsExpr, Ikarus::Util::VectorizeWithVoigt,    \
+                           Ikarus::Util::MatricizeWithVoigt)
+
+#define REGISTER_SIMPLE_SYMMETRIC_RESULTTYPE(structName, rowsExpr, ColsExpr, strainlike) \
```
Why can you specify `voigt` for simple but nur non-simple? @rath3t

rath3t replied:
Because Voigt only works for fixed size matrices at  compile time

henrij22 replied:
But atm 
```cpp
  #define REGISTER_SYMMETRIC_RESULTTYPE(structName, rowsExpr, ColsExpr, MaxRowsExpr, MaxColsExpr) \
    REGISTER_RESULTTYPE_IMPL(structName, rowsExpr, ColsExpr, Ikarus::Util::VectorizeWithVoigt,    \
                             Ikarus::Util::MatricizeWithVoigt)
```
uses `VectorizeWithVoigt`

henrij22 replied:
I think `REGISTER_SYMMETRIC_RESULTTYP` doens't really makes sense, because if `VectorizeWithVoigt` needs a compile-time matrix, then `maxRows` and `maxCols` doens't matter right?

rath3t replied:
Yes this function should be removed.

henrij22 committed:
Fixes from code review

fixies for testies

fix segfault

Fix Python

puuush

docstrings

henrij22 committed:
implement python registerCalculateAt

+ fixes from code review

henrij22 committed:
fix

henrij22 committed:
fix unused include

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,242 @@
+  }
+
+/**
+ * \brief Used to refister a symmetric ResultType with compile-time fixed rows and columns (uses Voigt notation)
```
typo

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,242 @@
+                           Ikarus::Impl::VectorizeWithVoigt<strainlike>, Ikarus::Impl::MatricizeWithVoigt<strainlike>)
+
+  /**
+   * \brief Used to refister a general ResultType with potentially dynamic size without reserved memeroy
```
typo

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,242 @@
+  REGISTER_RESULTTYPE_IMPL(resultTypeName, rowsExpr, colsExpr, Ikarus::Impl::VectorizeGeneric, \
+                           Ikarus::Impl::MatricizeGeneric)
+  /**
+   * \brief Used to refister a general ResultType with potentially dynamic size and defined maximum amount of
```
same

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,242 @@
+  REGISTER_RESULTTYPE_IMPL(resultTypeName, rowsExpr, colsExpr, MaxRowsExpr, MaxColsExpr,           \
+                           Ikarus::Impl::VectorizeGeneric, Ikarus::Impl::MatricizeGeneric)
+/**
+ * \brief Used to refister a general ResultType with potentially dynamic size, with reserved memory according to
```
same

rath3t commented on ikarus/python/finiteelements/registerelement.hh:
```diff
@@ -120,17 +120,35 @@ void registerElement(pybind11::handle scope, pybind11::class_<FE, options...> cl
+        Eigen::VectorXd result;
+        bool success = false;
+        Dune::Hybrid::forEach(restultTypesTuple, [&](auto i) {
+          using RT = typename std::invoke_result_t<decltype(i)>;
```
What is `decltype(i)` returning that was wrong? i would have assumed `std::remove_cvref_t<decltype(i)>` would be enough? Why the strange construction with invoke?

henrij22 replied:
I don't know why, but is a function pointer returning the resulttype. 
``` cpp
Ikarus::ResultType::linearStress<double, 1, 1> (*)()
```

rath3t commented on ikarus/python/finiteelements/registerelement.hh:
```diff
@@ -120,17 +120,35 @@ void registerElement(pybind11::handle scope, pybind11::class_<FE, options...> cl
+        bool success = false;
+        Dune::Hybrid::forEach(restultTypesTuple, [&](auto i) {
+          using RT = typename std::invoke_result_t<decltype(i)>;
+          if (resType == toString<RT::template Rebind>()) {
```
why not using toString(i)?

rath3t commented on ikarus/finiteelements/feresulttypes.hh:
```diff
@@ -0,0 +1,242 @@
+{
+  Vector,
+  Matrix,
+  Scalar
```
I would remove scalar here. This is actually not relly supported anyway for the rest of the code, is it?

henrij22 replied:
Nah you are right, I think for Scalar it doenst matter if its matrix or vector

rath3t commented on ikarus/utils/functionhelper.hh:
```diff
@@ -37,5 +39,23 @@ void obtainLagrangeNodePositions(const LV& localView,
+ * \return view over the position of the subenties
+ */
+template <typename FE>
+auto referenceElementSubEntitiyPositions(FE& fe, const int codim = FE::Traits::mydim) {
```
I think this default is strange. I would prefer no default. If you really want a short usage you can define an wrapper for the old function name:

```cpp
template <typename FE>
auto referenceElementVertexPositions(FE& fe) {
return referenceElementSubEntitiyPositions(fe, FE::Traits::mydim)
}
```

henrij22 replied:
Ah there is also a typo

rath3t commented on ikarus/utils/functionhelper.hh:
```diff
@@ -37,5 +39,23 @@ void obtainLagrangeNodePositions(const LV& localView,
+ * \return view over the position of the subenties
+ */
+template <typename FE>
+auto referenceElementSubEntitiyPositions(FE& fe, const int codim = FE::Traits::mydim) {
```
The const in the function parameter is also not nice style. It should by just int

rath3t reviewed (changes requested):
See comments

henrij22 committed:
code review fixes

henrij22 committed:
fix md format

henrij22 committed:
Fix eas

henrij22 committed:
fix test

rath3t commented on ikarus/python/finiteelements/nonlinearelastic.hh:
```diff
@@ -101,16 +103,8 @@ void registerNonLinearElastic(pybind11::handle scope, pybind11::class_<NonLinear
-          DUNE_THROW(Dune::NotImplemented, "Nonlinear-elastic element only supports PK2 stress as result.");
-      },
-      pybind11::arg("feRequirements"), pybind11::arg("local"), pybind11::arg("resultType"));
+  auto resultTypes = Dune::makeTupleVector(makeRT<ResultType::PK2Stress>);
```
Here the call operator is missing. This should be `makeRT<ResultType::PK2Stress>()`

henrij22 committed:
fix python
